Sprint 13 / Homework / Add player activity

herrfreeman opened the pull request.

herrfreeman committed:
Sprint 13 / Homework / Add player activity

faritowich commented on app/src/main/java/com/practicum/playlistmaker/Track.kt:
```diff
@@ -5,5 +5,14 @@ data class Track(
+
+) {
+    companion object {
+        const val EXTRAS_KEY = "TRACK_CLASS"
```
Не уверен, что это хорошее решение. Эта моделька будет изпользоваться много где в проекте, а EXTRAS_KEY только для передачи данных из SearchActivity в PlayerActivity. Лучше где-то там её и оставить

faritowich commented:
![image](https://github.com/herrfreeman/playlist-maker/assets/64851346/75dc807b-f813-4a63-9367-0cba7fcbd27b)

Полосатый экран и нет закруглений

faritowich commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,63 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_player)
+
+        track = gson.fromJson(intent.extras?.getString(Track.EXTRAS_KEY) ?: "", Track::class.java)
```
Текущая реализация через передачу строки getString небезопасная, так как если сейчас случайно передать из SearchActivity любой другой класс, то всё будет работать до  момента перехода на этот экран, после чего приложение упадет в рантайме. На самом деле можно передавать сам класс Track. Достаточно сделать его Serializable/Parcelable. Рекомендую почитать про это документацию, обычно делается именно так.

herrfreeman replied:
Изначально хотел сделать serializable, но SDK предлагает увеличить версию min sdk до 33, тем самым снизив количество поддерживаемых устройств, либо добавить какуюто непонятную аннотацию. Сделать serializable без этого нельзя?

faritowich commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,63 @@
+
+
+class PlayerActivity : AppCompatActivity() {
+    private val gson = Gson()
```
Как будто пока необязательно прям глобальную переменную в классе для этого объявлять

faritowich commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,63 @@
+        track = gson.fromJson(intent.extras?.getString(Track.EXTRAS_KEY) ?: "", Track::class.java)
+
+        val trackName = findViewById<TextView>(R.id.track_name)
+        val groupName = findViewById<TextView>(R.id.group_name)
```
Рекомендую почитать про библиотеку ViewBinding от гугла и использовать её - она используется повсеместно

faritowich commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,63 @@
+        trackCountry.text = track.country
+
+        Glide.with(this)
+            .load(track.artworkUrl100.replaceAfterLast('/',"512x512bb.jpg"))
```
Можно вынести логику по преобразованию ссылки artworkUrl100 в другое разрешение в сам класс Track, создав там соответствующую переменную, например так:
```
    val artworkUrl512 
        get() = artworkUrl100.replaceAfterLast('/', "512x512bb.jpg")
```

faritowich commented on app/src/main/res/layout/activity_player.xml:
```diff
@@ -0,0 +1,250 @@
+    <TextView
+        android:id="@+id/track_duration_value"
+        style="@style/Widget.PlayListMaker.Player.Value"
+        android:text="5:35"
```
Здесь и далее: для удобства визуального восприятия верстки в TextView удобно использовать атрибут `tools:text="Track Name"`
Он позволит указать текста в лейауте только для разработчика, при этом в приложении их не будет. 
Тогда студия не будет ругаться и подсвечивать текст желтым
Чтобы использовать его, надо добавить в корневой лейаут строку:
    `xmlns:tools="http://schemas.android.com/tools"`

herrfreeman committed:
Sprint 13 / Homework / Fix 1
Rebuild to use bindings, made Track class Serializable, fixed design issues: not rounded corners, strapes
